### ISSUE TITLE
Ability to exclude some directories/files from stories

### ISSUE THREAD

**Issue by pelotom** [storybookjs/storybook#11181]
The new `main.js` format for specifying stories requires specifying an array of globs, e.g.

```js
module.exports = {
  stories: ['../../src/**/*.@(story|stories).tsx'],
};
```

With this approach it's not clear how I can exclude certain things from being included. When the stories were configured within the browser environment I was able to use `require.context` with a regex, e.g.

```js
configure(
  require.context('../../src', true, /(?<!\/some\/forbidden\/directory.*)\.(story|stories)\.tsx?$/),
  module,
);
```

I could walk the directory structure in node and just provide a large array of fully resolved file paths, but it'd be nice to just be able to provide a regex instead of a glob. Or is there a way to do this with globs that I'm missing?

**Comment by shilman** [comment]
Haven't tried this, but possible solution here: 
 https://github.com/prettier/prettier/issues/1358#issuecomment-295669193

**Comment by pelotom** [comment]
I can't seem to make that work, it ends up excluding all files.

**Comment by DHedgecock** [comment]
Is there anyway to pass options for `glob` in `main.js`? From the `glob` docs:

>>> Previously, this module let you mark a pattern as a "comment" if it started with a # character, or a "negated" pattern if it started with a ! character.

These options were deprecated in version 5, and removed in version 6.

To specify things that should not match, use the ignore option.

**Comment by shilman** [comment]
cc @yannbf @ndelangen

**Comment by yannbf** [comment]
I don't think there's a mechanism in place for that, and I tried tweaking a little [here](https://github.com/storybookjs/storybook/blob/29a13f6657cb5c4d9828431008b2a0a728c8cade/lib/core/src/server/preview/to-require-context.ts#L36) but it seems like passing a `ignore` option to micromatch doesn't alter the resulting regex. Would be nice to have an `exclude` property like tsconfig has for instance, but unfortunately we don't have that. 

@pelotom I'm not sure about your use case, but the use case I see about using exclude would be maybe if I'm developing `CheckoutFeature` and I only want Storybook to bring those stories because maybe the project has like 500 stories. But in this scenario, rather than excluding, I would make my glob more specific and I would achieve that with the existing implementation of Storybook. Does that help somehow?

**Comment by pelotom** [comment]
> @pelotom I'm not sure about your use case, but the use case I see about using exclude would be maybe if I'm developing `CheckoutFeature` and I only want Storybook to bring those stories because maybe the project has like 500 stories. But in this scenario, rather than excluding, I would make my glob more specific and I would achieve that with the existing implementation of Storybook. Does that help somehow?

My use case is that I have one particular directory of stories nested deep within my repo which I want to exclude during visual regression testing in CI. It's not practical to positively specify all the things I do want to include while omitting this one particular deeply-nested directory, if that makes sense. My short term workaround is just to delete the directory in the CI script before building the storybook, but it'd be nice to have an `exclude` option.

**Comment by shilman** [comment]
@pelotom One workaround might be to specify the `disable` parameter for the components in that directory. This is available for both Storyshots and Chromatic, and may be available for whatever tool you're using.

```
export default {
  title: 'path/to/MyComponent',
  component: MyComponent,
  parameters: {
    chromatic: { disable: true },
  }
}
// stories here ...
```

**Comment by pelotom** [comment]
Ah, I am using Chromatic, thanks for the tip!

**Comment by artaommahe** [comment]
Faced same issue. We have a big library with a lot of submodules and their stories. Some submodules have their own storybook configuration due to specific code/environment. We need to exclude these submodules stories from the main storybook and cant find a way to do it with the new pattern format 😕

**Comment by shilman** [comment]
@artaommahe have you looked at storybook composition? https://medium.com/storybookjs/storybook-composition-af0da9084fba

**Comment by artaommahe** [comment]
how does it exclude dirs from specific storybook? e.x. this folders structure
```
.storybook/
layout/
lesson/
user/
widgets/
  chat/
    .storybook/
  switcher/
    .storybook/
```
we need layout/lesson/user stories, but not widgets one for the main storybook

**Comment by shilman** [comment]
you could just include `layout`, `lesson`, `user` and then compose the storybooks from chat & switcher if you wanted a unified view including storybooks with different configurations

**Comment by artaommahe** [comment]
hm, yu're right, just list all folders.. so there is no other way, no excluding in current glob implementation?

**Comment by ndelangen** [comment]
You're not required to specify globs!
You can specify an array of files.

```js
module.exports = {
  stories: async (list) => [...list, ...findFilesUsingWhateverMethodYouLikeAndReturnAnArray()],
};
```

**Comment by doublejosh** [comment]
Same need for excluding SCSS files from rebuilding Storybook... when using a parallel build and injecting via `.preview.js`

Otherwise, storybook rebuilds BEFORE the freshly compiled css is ready.

This seems like a common workflow (it's mentioned in docs)... is there a workaround/solution for that already?

**Comment by Berci-Ionut** [comment]
I also kind of need this. Is way harder to keep everything in sync by just adding all the new folders everytime. Is kind of redundant and hard to do. This option would be extremely  useful or at least a way to do it with regex.

**Comment by Domiii** [comment]
For anyone still struggling: You can in fact exclude stories without custom story resolver functions.

First of all, these are the relevant resources (as I also explained in #10153):

1. The official documentation on this is now [here](https://storybook.js.org/docs/react/configure/overview#configure-story-loading)
2. As explained on that page, the glob syntax is that of [micromatch](https://github.com/micromatch/micromatch#extended-globbing) (and not that of [glob](https://www.npmjs.com/package/glob)).
3. Here is the storybook code that parses and resolves stories: https://github.com/storybookjs/storybook/blob/master/lib/core-common/src/utils/to-require-context.ts#L31

Since [micromatch](https://github.com/micromatch/micromatch#extended-globbing) is used, you can simply add negative patterns using `!`.

**Comment by kelly-tock** [comment]
I am still not getting this to work, here's my current config:
```
../stories/**/*.stories.@(js|jsx|ts|tsx)
```

I basically want to exclude one folder inside of stories to keep it in source code for a while. I could move it out but don't really want to do that. 

I tried this:

```
'../stories!(/deprecated)/**/*.stories.tsx',
```

**Comment by Domiii** [comment]
@kelly-tock I would recommend installing `micromatch` locally to test your patterns. Yours looks good to me, but might be an issue with how they match things.

**Comment by dannyhw** [comment]
In case it's useful for anyone using node's glob with the ignore option worked well for me.

```
const glob = require('glob');
const path = require('path');
const appDirectory = path.resolve(__dirname, '../');
const getStories = () =>
  glob.sync(`${appDirectory}/stories/**/*.stories.@(js|jsx|ts|tsx|mdx)`, {
    ignore: `${appDirectory}/stories/**/*.native.stories.@(js|jsx|ts|tsx|mdx)`,
  });
module.exports = {
  stories: async list => [...list, ...getStories()],
}
```

**Comment by sneko** [comment]
On my side, @dannyhw the disadvantage is you have to stop/restart your Storybook to take in account new story files. While developing it's a bit annoying.

I recommend using in the array the standard way of Storybook with wildcards, and for a folder where you need to exclude something, you calculate fixed paths with `glob.sync` for example.

It's not perfect until Storybook will eventually manage excluding paths natively.

**Comment by dannyhw** [comment]
@sneko yeah you're right it doesn't work for new files, its more of a workaround. Though with the way require.context is used for importing stories I wasn't able to find another way to solve this at the time.

**Comment by sneko** [comment]
@dannyhw still no other good way same 1 year later. In my case I have `pnpm` in a monorepo, and storybook looks into `node_modules`, but since `pnpm` works natively with symlinks, it's like an infinite loop hell.

`glob` helped during a time, but I was tired of restarting Storybook at each new file, so I just moved all my "logic" code in big "src/" folders in each of my monorepo packages, like that I tell Storybook to not target `/package1` but instead `/package1/src`, so no more circular loops with `node_modules` folders.

**Comment by D1no** [comment]
In a pnpm workspace with storybook on vite, the `!(/node_modules)` ignore glob doesn't work. 

```js
const config: StorybookConfig = {
  stories: [
  // (...)
    "../ui/**/*.stories.@(js|jsx|ts|tsx)",
```
 * recursively sources stories from the ui library when components within cross pnpm symlink each other deep down

```js
    "../ui/**!(/node_modules)/**/*.stories.@(js|jsx|ts|tsx)",
```
 * trying to exclude any node_modules folder down the tree = no luck, even after many attempts and reading [micromatch](https://github.com/micromatch/micromatch#extended-globbing) docs.

Probably related to
https://github.com/storybookjs/storybook/issues/15913 (need for ignore options)

**Comment by andrejilderda** [comment]
@D1no I had a very similar problem within a similar project setup which I managed to fix by:

```js
  stories: [
    "../../**!(node_modules)/**!(node_modules)/*.mdx",
    "../../**!(node_modules)/**!(node_modules)/*.stories.@(js|jsx|ts|tsx)",
  ],
```
See: https://github.com/andrejilderda/desktop-ui/commit/0993a07ec7a1af0a72f88d090251f0d62f70a8f1

Also note that Storybook is not using micromatch but [globby](https://github.com/sindresorhus/globby/) as a globbing library (see [here](https://github.com/storybookjs/storybook/blob/3f7c1e1a19fb19e07f9c07b81c48192620157b53/code/lib/core-server/src/utils/StoryIndexGenerator.ts#LL4C3-L4C3)).

**Comment by D1no** [comment]
Thank you very much @andrejilderda ! Due to your insight I was able to add it to the following issue: https://github.com/storybookjs/storybook/issues/21399

Seems like it is flaky / not really user friendly but it works! [Reference](https://github.com/D1no/reproduction-storybook-symlinks-pnpm/blob/aa7b519855875d6269ec7a1f63f4b9d401a5297e/tools/.storybook/main.ts#L55-L69)
```ts
  case "4_proper_workspace_stories_trying_to_avoid_recursion_with_glob":
    configStories = [
      /**
       * Since storybook uses globby under the hood, trying to add a negative
       * glob for nested node_modules via "**!(node_modules)/**!(node_modules)".
       * -> Insight from https://github.com/storybookjs/storybook/issues/11181#issuecomment-1474347928
       *    though sometimes just "**!(node_modules)" seems to work, when there
       *    are stories directly on next level folder. The following wont work:
       *    "../node_modules/@repo/views/**!(node_modules)/**!(node_modules)/*.stories.@(js|jsx|ts|tsx)"
       *    This does:
       *    "../node_modules/@repo/views/**!(node_modules)/*.stories.@(js|jsx|ts|tsx)"
       */
      "../node_modules/@repo/**!(node_modules)/**!(node_modules)/*.stories.@(js|jsx|ts|tsx)",
    ];
    break;
```

**Comment by astriskit** [comment]
just to add the point - tried the above steps of glob pattern - `**(!)` including `!<dir>` too; but couldn't get to intended ignore; hence went old-school by having two stories array, one for the build-time and one for the dev-time. And the stories, which are to be separated out are configured aptly in both stories. Basically one of the array has all the story-file extensions included, including the build-time-non-inclusive files (used *.d-stories.tsx) and the other array has all except the .d-stories.tsx files.

Btw, the use-case was to skip some stories while building. And the storybook setup is 7-rc3.

**Comment by oemer-sellwerk** [comment]
As stated before,  storybook is using `globby`, but adjust it's behaviour internally. My wild guess is, that storybook prepends `.storybook/` to every glob pattern?

With a workaround, you can use `globby` just as it is documented: https://www.npmjs.com/package/globby

Here is how I include all `**/*.stories.@(js|jsx|ts|tsx)`, excluding `node_modules`:

```ts
import type { StorybookConfig } from "@storybook/vue3-vite"
// This is already installed by storybook
import globby from "globby"

const config: StorybookConfig = {
  // globby is working from the root, where "yarn storybook" was called. 
  // To make it work with storybook, we have to prepend "../" to every path that was found
  stories: globby.sync([`**/*.stories.@(js|jsx|ts|tsx)`, "!**/node_modules/**/*"]).map((path) => "../" + path),
  // ...
}
export default config

```
None of the other workarounds worked for me.

**Comment by Haschtl** [comment]
> As stated before, storybook is using `globby`, but adjust it's behaviour internally. My wild guess is, that storybook prepends `.storybook/` to every glob pattern?
> 
> With a workaround, you can use `globby` just as it is documented: https://www.npmjs.com/package/globby
> 
> Here is how I include all `**/*.stories.@(js|jsx|ts|tsx)`, excluding `node_modules`:
> 
> ```ts
> import type { StorybookConfig } from "@storybook/vue3-vite"
> // This is already installed by storybook
> import globby from "globby"
> 
> const config: StorybookConfig = {
>   // globby is working from the root, where "yarn storybook" was called. 
>   // To make it work with storybooks internal logic, we have to change the cwd 
>   stories: globby.sync([`../**/*.stories.@(js|jsx|ts|tsx)`, "!../**/node_modules/**/*"], , { cwd: "./.storybook" }),
>   // ...
> }
> export default config
> ```
> 
> None of the other workarounds worked for me.

Thanks for this solution! 
Since globby@12 the `globby.sync` was replaced with `globbySync`.
Here is the snippet for globby@12

```js
import type { StorybookConfig } from "@storybook/vue3-vite"
// This is already installed by storybook
import { globbySync } from "globby"

const config: StorybookConfig = {
  // globby is working from the root, where "yarn storybook" was called. 
  // To make it work with storybooks internal logic, we have to change the cwd 
  stories: globbySync([`../**/*.stories.@(js|jsx|ts|tsx)`, "!../**/node_modules/**/*"], , { cwd: "./.storybook" }),
  // ...
}
export default config
```

**Comment by oemer-aran** [comment]
> > As stated before, storybook is using `globby`, but adjust it's behaviour internally. My wild guess is, that storybook prepends `.storybook/` to every glob pattern?
> > With a workaround, you can use `globby` just as it is documented: https://www.npmjs.com/package/globby
> > Here is how I include all `**/*.stories.@(js|jsx|ts|tsx)`, excluding `node_modules`:
> > ```ts
> > import type { StorybookConfig } from "@storybook/vue3-vite"
> > // This is already installed by storybook
> > import globby from "globby"
> > 
> > const config: StorybookConfig = {
> >   // globby is working from the root, where "yarn storybook" was called. 
> >   // To make it work with storybooks internal logic, we have to change the cwd 
> >   stories: globby.sync([`../**/*.stories.@(js|jsx|ts|tsx)`, "!../**/node_modules/**/*"], , { cwd: "./.storybook" }),
> >   // ...
> > }
> > export default config
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > None of the other workarounds worked for me.
> 
> Thanks for this solution! Since globby@12 the `globby.sync` was replaced with `globbySync`. Here is the snippet for globby@12
> 
> ```js
> import type { StorybookConfig } from "@storybook/vue3-vite"
> // This is already installed by storybook
> import { globbySync } from "globby"
> 
> const config: StorybookConfig = {
>   // globby is working from the root, where "yarn storybook" was called. 
>   // To make it work with storybooks internal logic, we have to change the cwd 
>   stories: globbySync([`../**/*.stories.@(js|jsx|ts|tsx)`, "!../**/node_modules/**/*"], , { cwd: "./.storybook" }),
>   // ...
> }
> export default config
> ```

Unfortunately, this breaks HMR for new files.. If you add a new `.stories.ts` it will not be recognized until restart. This makes sense because `globbySync`/`globby.sync` just returns an array of all the file paths. The new file will never be added to this array after it already started.

This issue has to be solved internally by the Storybook team to support globby patterns as expected.

**Comment by sneko** [comment]
I'm in a monorepo and some imports are done for modules, so I need to mix both worlds. The easiest solution for me was:
```ts
# .storybook/main.ts

const environmentVariables: Record<string, string> = {
  LOL: 'true',
};

export const config: StorybookConfig = {
    ...,
    env: (config) => ({
        ...config,
        ...environmentVariables,
    }),
    viteFinal: {
        define: {
            'process.env': environmentVariables,
        },
    },
};
```

**Comment by pm0u** [comment]
Just want to mention that it seems (according to the docs) Storybook now uses (picomatch)[https://github.com/micromatch/picomatch] which may be missing some features from micromatch (why i was left scratching my head)

**Comment by prashanth-cn** [comment]
> As stated before, storybook is using `globby`, but adjust it's behaviour internally. My wild guess is, that storybook prepends `.storybook/` to every glob pattern?
> 
> With a workaround, you can use `globby` just as it is documented: https://www.npmjs.com/package/globby
> 
> Here is how I include all `**/*.stories.@(js|jsx|ts|tsx)`, excluding `node_modules`:
> 
> ```ts
> import type { StorybookConfig } from "@storybook/vue3-vite"
> // This is already installed by storybook
> import globby from "globby"
> 
> const config: StorybookConfig = {
>   // globby is working from the root, where "yarn storybook" was called. 
>   // To make it work with storybooks internal logic, we have to change the cwd 
>   stories: globby.sync([`../**/*.stories.@(js|jsx|ts|tsx)`, "!../**/node_modules/**/*"], , { cwd: "./.storybook" }),
>   // ...
> }
> export default config
> ```
> 
> None of the other workarounds worked for me.

`  core: {
    builder: '@storybook/builder-webpack5',
  },`

pointing to `@storybook/builder-webpack5` in main.js fixed the HMR for me

**Comment by eportet** [comment]
> Just want to mention that it seems (according to the docs) Storybook now uses [picomatch](https://github.com/micromatch/picomatch) which may be missing some features from micromatch (why i was left scratching my head)

Thanks to @pm0u. Reading the documentation for picomatch I saw that their negation operator works this way:

> !(pattern): Match _**anything**_ but pattern

So ended up writing my storybook config like so:
```typescript
// main.ts
const config: StorybookConfig = {
  // ...
  stories: ['../src/!(ignoreFolder)/**/*.stories.{ts,tsx}', '../src/**/*.mdx'],
  // ...
}
```

**Comment by ingvaldlorentzen** [comment]
Can be kinda tricky to get this right, so thought I would share my solution with anyone coming across this. 
```
stories: [
    '../apps/**/!(tests)/*.mdx',
    '../apps/**/!(tests)/*.stories.@(ts|tsx)',
    '../packages/**/!(tests)/*.mdx',
    '../packages/**/!(tests)/*.stories.@(ts|tsx)',
  ],
```

**Comment by Fox32** [comment]
> ```
> stories: [
>     '../apps/**/!(tests)/*.mdx',
>     '../apps/**/!(tests)/*.stories.@(ts|tsx)',
>     '../packages/**/!(tests)/*.mdx',
>     '../packages/**/!(tests)/*.stories.@(ts|tsx)',
>   ],
> ```

Anyone else having trouble to use this approach with the latest beta (`v10.1.0-beta.2`)? For me it feels like the negation pattern doesn't work anymore and I now have to include everything one by one.

**Comment by ACERpos** [comment]
Some additional detail I'd like to provide on this issue for anybody struggling with it. In `10.4.1` we have found that issues regarding recursive story loading in monorepos due to matching stories in nested `node_modules` directories is largely gone.

However there seem to be two incompatible follow-on issues using the startup globbing workaround some propose above:

1. When pre-matching and passing a resolved list of story files to `stories`, both the initial startup time **and memory usage** of the Storybook server are dramatically increased. For us this began causing OOM exceptions.
2. When using the standard configuration as an array of glob matchers, we had issues with CI timing out when initialising the Storybook. This resulted in "Failed to connect to the browser session" errors with Playwright.

Thus we ended up needing to use a mix of both approaches depending on the execution environment. We updated our test script to set an environment variable:

```
"test-storybook": "cross-env STORYBOOK_CI=1 vitest --project=storybook --coverage"
```

and update our Storybook `main.ts` to conditionally pre-match stories as needed:

```ts
let stories = [
  // "standard" configuration you would expect to work
  "../packages/**/*.stories.@(js|jsx|mjs|ts|tsx)",
];

if (process.env.STORYBOOK_CI) {
  // in CI mode, statically parse all stories initially at startup to prevent
  // long hang when initing test runner, causing browser connection error
  stories = globbySync([...stories, "!../**/node_modules/**/*"], { cwd: "./.storybook" });
} else {
  // in devserver mode, use standard matchers to boot quicker & dynamically load tests on demand
}

const config: StorybookConfig = {
  stories,
  // etc
};
export default config;
```
